Rework placeholder count inspection (fixes #2188)

dlsniper opened the pull request:
This matches the inspection with what `go vet` does. The problem is that `go vet` currently doesn't match the language specs / what the Go compiler does, thus the failing tests.
I'll work into getting this back in working order without changing the tests as the error is in the inspection logic not the test code.

zolotov commented:
@dlsniper thanks for PR. This is really great that you repeated whole `govet` logic. The only thing is that it's better to write Java in Java style, not in Go. I'll comment some snippets with inappropriate go-style in java.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+      scanNum();
+      if (nBytes == format.length() || nBytes == start || format.charAt(nBytes) != ']') {
+        String message = "Illegal syntax for <code>#ref</code> argument index";
+        myHolder.registerProblem(myPlaceholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
```
This is strange that `state`-class reports something to problems holder. Actually it shouldn't do anything besides storing data. But ok, at least it should not interact to outer world.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+    }
+
+    private void scanNum() {
+      for (; nBytes < format.length(); nBytes++) {
```
we have `while` loop in java

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+    private void scanNum() {
+      for (; nBytes < format.length(); nBytes++) {
+        if (format.charAt(nBytes) < '0' || '9' < format.charAt(nBytes)) {
+          return;
```
Character#isDigit?

dlsniper replied:
From what I understand in the method documentation, Character#isDigit will match also non-Latin digits. Go is explicit at supporting only ` 0-9 ` digits for everything here so we could even add a "Number is not a valid Latin-1 digit" error here. What do you think?

zolotov replied:
You're right. `StringUtil#isDecimalDigit` then

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+import static com.goide.GoConstants.TESTING_PATH;
+import static com.goide.inspections.GoPlaceholderChecker.PrintfArgType.*;
+
+public class GoPlaceholderChecker {
```
Fill nullability

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+  }
+
+  @Nullable
+  private static String resolve(@NotNull GoExpression argument) {
```
I don't understand the ordering of methods. I read three methods and didn't understand why I did this

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+  }
+
+  private static boolean isStringPlaceholder(@NotNull GoExpression argument) {
+    GoType goType = argument.getGoType(null);
```
see GoTypeUtil

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+    public char verb;             // the format verb: 'd' for "%d"
+    public String format;         // the full format directive from % through verb, "%.3d".
+    public String flags;          // the list of # + etc.
+    public List<Integer> argNums; // the successive argument numbers that are consumed, adjusted to refer to actual arg in call
```
it's always initialized in `parsePrintfVerb` and it's always `new ArrayList`. Do you have any reason to not to declare it as final

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+    }
+  }
+
+  @NotNull private static List<PrintVerb> printVerbs = ContainerUtil.newArrayList(
```
why not final? are you going to reassign it?

why not enum? are you going to change inner values?

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+  }
+
+  private class FormatState {
+    public char verb;             // the format verb: 'd' for "%d"
```
Why all fields are public? Who is responsible for changing state of `state`? What `FormatState` is it exactly, parser or state? If parser, it shouldn't expose all it's state to anyone and shouldn't allow to change it. If state, then it shouldn't parse anything.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,497 @@
+    PrintVerb v = null;
+    boolean found = false;
+
+    for (int i = 0; i < printVerbs.size(); i++) {
```
foreach

zolotov replied:
`found` variable is completely redundant: if you cannot find `verb` the `v` variable should be null.
whole loop looks silly. just replace printVerbs with any map.

dlsniper commented:
@zolotov thank you so much for the review. I was fir focused on getting it to run then passing the tests then make it correct Java. Your comments will help me out a lot, I'll try to fix everything today. Thank you!

dlsniper commented:
I'll also add the test data from `go vet` as soon as our tests are fixed and the code is written according to Java style. For now, `go vet` still has inferior matching of our current solution so I'll also contribute back to Go the fix for that once it's working.

dlsniper commented:
@zolotov I've updated this. I've changed the order of methods as well, but not sure if that has improved the readability or not. All the other items have been done.
I'm not sure if this is the best construct to solve the the print verbs definition 
```
private enum PrintVerb {
    Percent(Couple.of('%', Couple.of("", 0))),
...
```
so maybe that's something to look after. It could be simplified for now as we don't use the type mask yet (needed in the type assertion between parameter type and placeholder type).

Once this passes your review, I'll proceed with adding the tests from `go vet` and checking if any other changes need to be done.

Thank you.

zolotov commented:
Actually I don't understand why do you need Couple there. 

1. `Couple` is shorthand for `Pair` and should be used when both object have the same type
2. It's supposed to use static types anywhere you can in Java
3. Why don't just 
```
private enum PrintVerb {
PERCENT('%', "", 0),
 b('b', " -+.0", argInt.getValue() | argFloat.getValue() | argComplex.getValue())
...
```

with descriptive argument names, types and so on.

dlsniper commented:
> 3. Why don't just

Oh, mostly because I didn't knew I can do that in Java and the enum docs I've seen weren't mentioning that. I'll change it right now. Thank you.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,502 @@
+    argPointer(7),
+    anyType(-1);
+
+    private int myVal;
```
The comment says
> // The types of expressions a printf verb accepts. It is a bitmask.

The first part is presented in enum's name. The second you can show by renaming `myVal` variable and `getValue` method. Comment is redundant.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,502 @@
+    Pair.pair("sprintf", 0),
+    Pair.pair("sscanf", 1));
+
+  public static final List<String> PRINT_FUNCTIONS = ContainerUtil.newArrayList(
```
1. the field is used only in `GoPlaceholderCountInspection`, so why it's here and public?
2. if the list is used for `contains` operation only it shouldn't be a list.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,502 @@
+  }
+
+  private enum PrintVerb {
+    Percent(Couple.of('%', Couple.of("", 0))),
```
Since you don't depend on enum's fields names it makes sense to make them more descriptive

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,502 @@
+  private FormatState parsePrintfVerb(@NotNull GoExpression placeholder, @NotNull String format, int firstArg, int argNum) {
+    FormatState state = new FormatState(format, argNum, firstArg);
+
+    // There may be flags
```
comments about flags before invoking `parseFlags` method looks redundant

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,502 @@
+
+    if (!hasPlaceholder(placeholderText) && arguments.size() > placeholderPosition) {
+      String message = "Value used for formatting text does not appear to contain a placeholder";
+      myProblemsHolder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
```
Still strange dependency on problems holder here. You have a parser which just parses some string.
First of all it's weird that the parser knows about problemsHolder. In the second, it means that you can parse printf placeholder only in inspections, so no completion support in future. In the third, what is the real responsibility of GoPlaceholderChecker? If it's a Checker it should check and return the result (success or not). If it's parser it should parse and return parsed representation of printf argument (e.g. a Map). If it's inspection, it should visit elements and create problem descriptors for some of them. You're trying to create a class that knows everything and does all these three things. It's hard to read, hard to modify and hard to reuse. Try to keep single responsibility for each class and method.

dlsniper replied:
I have the same opinion on this. I'll try to organize the code better and split its concerns but this might take a bit. Thank you.

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -162,23 +40,18 @@ public void visitCallExpr(@NotNull GoCallExpr o) {
-        int placeholdersCount = getPlaceholders(placeholder);
-        if (placeholdersCount == -1 || placeholdersCount == parameterCount) return;
+        String functionName = ((GoFunctionOrMethodDeclaration)resolved).getName();
+        if (functionName == null) return;
```
toLowerCase return null for null argument, no needed to check function name here, check it later and avoid variable reassignments

dlsniper commented:
I've split up all the things and our tests are passing (in the process I was able to add a few more cases previously unsupported). I hope the layout it's better now.
The next step is to fix the inconsistency in handling indexes as there's currently an off-by-one error in some cases (highlighted by the go vet tests).
I'll continue in the coming days with fixing this.
Finally, I've noticed that the performance tests are failing but I don't know what could be the cause for this (I am receiving emails from TeamCity about them).

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,456 @@
+  }
+
+  public static boolean isPrintingFunction(String functionName) {
+    return PRINTING_FUNCTIONS.contains(functionName);
```
avoid `contains` on lists

dlsniper replied:
I've tried `indexOf()` but the IDEA will say I should use `contains()`. Should I use a simple `for` instead?

zolotov replied:
No, use Set instead

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,456 @@
+      MISSING_VERB_AT_END(1),
+      ARGUMENT_INDEX_NOT_NUMERIC(2);
+
+      private int state;
```
is this used anywhere?

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,456 @@
+  }
+
+  private static class FormatState {
+    private char verb;                  // the format verb: 'd' for "%d"
```
As I can see this field is not used much. It's compared to `%` and it's converted to Verb. So why it's not nullable Verb field?

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,263 @@
+    int maxArgsNum = 0;
+    for (Placeholder placeholder : placeholders) {
+      List<Integer> arguments = placeholder.getArguments();
+      if (arguments == null || arguments.isEmpty()) continue;
```
is arguments nullable?

zolotov commented on testData/highlighting/struct.go:
```diff
@@ -58,7 +58,7 @@ func main() {
     c := Car{4}
     Println("A Car has this many wheels:", c.wheelCount)
-    Println("A Car has this many wheels:", c.numberOfWheels)
+    Println("A Car has this many wheels:", c.numberOfWheels())
```
both lines are identical, is this intentionally?

zolotov commented on src/com/goide/psi/impl/GoTypeUtil.java:
```diff
@@ -41,6 +41,13 @@ private static boolean isArray(@Nullable GoType type) {
     GoType underlyingType = type != null ? type.getUnderlyingType() : null;
-    return underlyingType != null && underlyingType.textMatches("string") && GoPsiImplUtil.builtin(underlyingType);
+    return underlyingType != null &&
+           underlyingType.getText() != null &&
```
?

dlsniper replied:
That's to prevent a NPE with cgo, I've got it from the `cgotest.go` file.

zolotov replied:
maybe it makes sense to implement `textMatches` or `getText` methods for CTypes here. Or maybe check GoCType explicitly like `underlyingType instanceof GoCType`. Anyway test is required.

dlsniper replied:
I've changed this to check if the type is not `instanceof GoCType` and added a test for this but it's still in the placeholder test suite, it isn't contained on its own suite. Should I do that instead?

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,263 @@
+
+    if (argNum == 0) {
+      String message = String.format("Index value [0] for <code>#ref</code>(\"%s\"); indexes start at 1", fmtPlaceholder.getPlaceholder());
+      holder.registerProblem(placeholder, message, ProblemHighlightType.GENERIC_ERROR_OR_WARNING);
```
the function looks like it answers to the question, not reports any problems

dlsniper commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,456 @@
+
+    // We always receive the text with double quotes at the beginning and at the end so remove them
+    if (value != null) {
+      value = value.substring(1, value.length() - 1);
```
Is there a way to avoid getting the string value double quoted? There'd at least one more place where I had to take this into account to do the check.

zolotov replied:
there is a method StringUtil for this. As for `GoStringLiteral` it's better to use `com.intellij.psi.ElementManipulators#getValueText`

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,456 @@
+    }
+    if (expression instanceof GoAddExpr) {
+      String sum = getValue((GoAddExpr)expression);
+      return sum.isEmpty() ? null : sum;
```
StringUtil.nullize

dlsniper commented:
Sorry I've disappeared, I've fixed this and now tests pass again.

There are a few todos left in this (some mostly because we don't have type checking just yet). A few others could be done later on as an improvement over both the existing solution and `go vet` itself.

The other main difference between what we'll support now and what `go vet` does is that we won't support arbitrary functions which match the name and signature with the any of the printing functions (e.g. we support `fmt.Printf()` but don't support `dlsniper.Printf()` whereas would support it).

As for the unused methods, I've left them for now to be able to easily use them for various quickfixes which could be applied to this inspection.

Finally I should squash this before merging so let me know if you think it's ok to do so.

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,440 @@
+    return state;
+  }
+
+  protected static boolean hasPlaceholder(@NotNull String formatString) {
```
inspection is the only client of this method, move the method to inspection

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,440 @@
+  }
+
+  protected static boolean hasPlaceholder(@NotNull String formatString) {
+    return StringUtil.contains(formatString, "%");
```
Use `containsChar`

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,440 @@
+  }
+
+  @Nullable
+  protected static String resolve(@NotNull GoExpression argument) {
```
inspection is the only client of this method, move the method to inspection

dlsniper replied:
I was thinking this would also be useful in other use cases of this without the refactoring to move it here then. But sure I can move it.

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,261 @@
+    if (GoTypeUtil.isString(firstArg.getGoType(null))) {
+      String firstArgText = GoPlaceholderChecker.resolve(firstArg);
+      if (firstArgText != null) {
+        if (GoPlaceholderChecker.hasPlaceholder(firstArgText)) {
```
Any string with `%` will be highlighted? Even `fmt.Println("10%")`?

dlsniper replied:
Unfortunately yes. At least that's the current behavior of `go vet`. Maybe we should make this smarter and call for this if the string actually contains a formatting directive?

zolotov replied:
Not sure but it looks strange. Well let's keep it in this way until somebody complain

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,261 @@
-      if (maxIndexCount < indexCount) {
-        maxIndexCount = indexCount;
+      else if (GoTypeUtil.isFunction(goType)) {
+        String message = "Argument <code>#ref</code> is not a function call";
```
add `#loc` at the end, try to run `Inspect code` action to see how it looks

dlsniper replied:
Unfortunately it appears as an empty quoted string, ` '' `.

zolotov replied:
It doesn't.

<img width="957" alt="screenshot 2016-06-15 15 00 53" src="https://cloud.githubusercontent.com/assets/140920/16078898/030e9732-330a-11e6-924b-a763571282f2.png">

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,261 @@
-  @Nullable
-  private static String resolve(@NotNull GoExpression argument) {
-    if (argument instanceof GoStringLiteral) return argument.getText();
+  private static void checkPrintf(
```
Usually we format it in this way:

```
private static void checkPrint(@NotNull ProblemsHolder holder,
                                 @NotNull GoCallExpr callExpr,
                                 @NotNull GoFunctionOrMethodDeclaration declaration) {
```

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,261 @@
+
+    if (argNum <= callArgsNum) return true;
+    if (GoPsiImplUtil.hasVariadic(callExpr.getArgumentList())) return false;
+    if (argNum <= callArgsNum) return true;
```
typo

zolotov commented:
Some cleanup is needed, but the PR looks really good to me so far, thank you.

dlsniper commented:
I'll let TeamCity check for anything broken and fix it tomorrow morning.

As for the two methods you've suggested moving, I've left them to be used in other places if needed (completion?) and avoid later refactoring and moving things around. If you think I should move them however, I'll do so.

Finally, I'll create issues for the TODOs present in the code once this is fixed.

 Thank you for your patience with this PR.

zolotov commented:
there is no reason to avoid move-methods refactoring. Actually, you won't avoid it. `resolve` method has nothing to do with placeholder checking and it has a bad name to be public, so as soon as you'll reuse it you'll have to move it in more appropriate place and likely rename. Until then it should be `private`.

Ok, let's keep `hasPlaceholder` in checker. Make it public then

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,440 @@
+    String knownFlags = "#0+- ";
+    StringBuilder flags = new StringBuilder(state.flags);
+    while (state.nBytes < state.format.length()) {
+      if (knownFlags.indexOf(state.format.charAt(state.nBytes)) != -1) {
```
StringUtil.containsChar

zolotov commented:
Also please fill nullability annotations.

Btw, what about unused methods in checker?

zolotov commented:
Please consider reporting about particular text range instead of whole string literal

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,256 @@
+                                   @NotNull GoExpression placeholder,
+                                   @NotNull List<Placeholder> placeholders) {
+
+    for (Placeholder p : placeholders) {
```
it's really hard to understand why while traversing `placeholders` you have `placeholder` variable which is not instance of `Placeholder` class. Unreadable, please rename it properly

dlsniper commented:
> Also please fill nullability annotations.

GoPsiImplUtil.java needs this but it will touch too many things and I'd rather to this outside of this PR otherwise I've ran `Infer nullity` from IDEA on all the other files. Does it matter if I'm still using IDEA 15 for this?

> Please consider reporting about particular text range instead of whole string literal

Oh wow, I've missed that feature, that's really cool! :heart: it!

I've moved and followed the rest of the feedback with regards to visibility and location for methods.

Finally, I've added a bunch more todos and a couple of missing tests.

Thank you.

zolotov commented:
> GoPsiImplUtil.java needs this but it will touch too many things 

Fill nullability annotations only for code you've added

zolotov commented on src/com/goide/inspections/GoPlaceholderCountInspection.java:
```diff
@@ -16,170 +16,321 @@
+        String argText = resolve(argument);
+        if (argText != null && argText.endsWith("\\n")) {
+          String message = "Function already ends with new line #loc";
+          TextRange range = TextRange.create(argText.length() - 1, argText.length() + 1);
```
TextRange.from(argText.length() - 1, 2)

dlsniper replied:
I think I'm doing something wrong, if I do that, then the test will fail: `ERROR: Invalid range specified: (5,2); `
What's strange is that TextRange seems to be 1 indexed not 0 indexed but I might be wrong.

zolotov replied:
TextRange is not indexed at all, it's just container for two integers :)

dlsniper commented:
> Fill nullability annotations only for code you've added 

For the single function I've added there it's already correct as for the other files in the PR I've already done it (or I'm missing something as `Infer nullity` doesn't do any further changes in the code base).

Thank you.

dlsniper commented:
I'll rebase it when it's ready to merge / reviewed if needed. Thanks.

zolotov commented:
@dlsniper I'm ready to merge this with minor fixes. What about todos you've placed in code?

zolotov commented:
@dlsniper also what's the mask in PrintVerb stands for? It's still unused

dlsniper commented:
For the todos I'll create the according issues, I plan to work on adding the issues after Thursday if that's ok for you. As for PrintVerb, that might be useful in when solving the various todos such as checking the type and see if we can suggest a better one than the current, if it doesn't make sense to have it there for the future (as it's easier to keep it than figure out later how to get it) we can remove it from now, up to you. Should I rebase this first as it has quite a few commits that might not be relevant? Thank you.

dlsniper commented:
To be clear, I'll on on all the existing inspections as well and if if there's a todo or a missing quickfix, determine what's needed and create the according ticket, not just for this particular issue.

dlsniper commented:
and rebese means squash, sorry

zolotov commented:
Ok, thanks, will merge it then.

> Should I rebase this first as it has quite a few commits that might not be relevant? 

As you wish, I can squash all of them into a single commit while merging if you want.

dlsniper commented:
Done, thank you!

dlsniper commented:
I'll add the todos / missing quick fixes for this when it's merged, to be consistent with the other inspections.

zolotov commented:
merged manually

zolotov commented on src/com/goide/inspections/GoPlaceholderChecker.java:
```diff
@@ -0,0 +1,386 @@
+    return true;
+  }
+
+  private static boolean parseNum(@NotNull FormatState state) {
```
why is this always return `true`? and why other code supposes the function can return `false`?

dlsniper replied:
I'll have to look at gofmt again, I thought it's strange as well.
